pkg/controller/service/servicecontroller_test.go

```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+
+	testCases := []struct {
+		testName   string
+		runTest    func()
+		expectedFn func() error
+	}{
+		{
+			testName: "Add",
+			runTest: func() {
+				cS := sC.getOrCreate("addTest")
+				cS.state = defaultExternalService()
+			},
+			expectedFn: func() error {
+				//There must be exactly one element
+				if len(sC.serviceMap) != 1 {
+					return fmt.Errorf("Expected=1 Obtained=%d", len(sC.serviceMap))
+				}
+				return nil
+			},
+		},
+		{
+			testName: "Del",
+			runTest: func() {
+				sC.delete("addTest")
+
+			},
+			expectedFn: func() error {
+				//Now it should have no element
+				if len(sC.serviceMap) != 0 {
+					return fmt.Errorf("Expected=0 Obtained=%d", len(sC.serviceMap))
+				}
+				return nil
+			},
+		},
+		{
+			testName: "SetandGet",
+			runTest: func() {
+				sC.set("addTest", &cachedService{state: defaultExternalService()})
+			},
+			expectedFn: func() error {
+				//Now it should have one element
+				Cs, bool := sC.get("addTest")
+				if !bool {
+					return fmt.Errorf("is Available Expected=true Obtained=%v", bool)
+				}
+				if Cs == nil {
+					return fmt.Errorf("CachedService expected:non-nil Obtained=nil")
+				}
+				return nil
+			},
+		},
+		{
+			testName: "ListKeys",
+			runTest: func() {
+				//Add one more entry here
+				sC.set("addTest1", &cachedService{state: defaultExternalService()})
+			},
+			expectedFn: func() error {
+				//It should have two elements
+				keys := sC.ListKeys()
+				if len(keys) != 2 {
+					return fmt.Errorf("Elementes Expected=2 Obtained=%v", len(keys))
+				}
+				if keys[0] != "addTest" || keys[1] != "addTest1" {
+					return fmt.Errorf("Keys do not match")
+				}
+				return nil
+			},
+		},
+		{
+			testName: "GetbyKeys",
+			runTest: func() {
+				//Do nothing
+			},
+			expectedFn: func() error {
+				//It should have two elements
+				srv, isKey, err := sC.GetByKey("addTest")
+				if srv == nil || isKey == false || err != nil {
+					return fmt.Errorf("Expected(non-nil, true, nil) Obtained(%v,%v,%v)", srv, isKey, err)
+				}
+				return nil
+			},
+		},
+		{
+			testName: "allServices",
+			runTest: func() {
+				//Do nothing
+			},
+			expectedFn: func() error {
+				//It should return two elements
+				srvArray := sC.allServices()
+				if len(srvArray) != 2 {
+					return fmt.Errorf("Expected(2) Obtained(%v)", len(srvArray))
+				}
+				return nil
+			},
+		},
+	}
+
+	for _, tst := range testCases {
+		tst.runTest()
+		if err := tst.expectedFn(); err != nil {
+			t.Errorf("%v returned %v", tst.testName, err)
+		}
+	}
+}
+
+//Test a utility functions as its not easy to unit test nodeSyncLoop directly
+func TestUtilityFunctions_nodeSlicesEqualForLB(t *testing.T) {
+	numNodes := 10
+	nArray := make([]*v1.Node, 10)
+
+	for i := 0; i < numNodes; i++ {
+		nArray[i] = &v1.Node{}
+		nArray[i].Name = fmt.Sprintf("node1")
+	}
+	if !nodeSlicesEqualForLB(nArray, nArray) {
+		t.Errorf("Expected=true Obtained=false")
```

Comment:
Could we also put the function name (nodeSlicesEqualForLB) in the error log to make it clearer?